lib/model/recent_senders.dart

```diff
@@ -0,0 +1,155 @@
+    (((topicSenders[streamId] ??= {})[topic] ??= {})
+      [senderId] ??= MessageIdTracker()).add(messageId);
+  }
+
+  void handleDeleteMessageEvent(DeleteMessageEvent event, Map<int, Message> cachedMessages) {
+    if (event.messageType != MessageType.stream) return;
+    final messagesByUserInStream = <(int, int), List<int>>{};
+    final messagesByUserInTopic = <(int, String, int), List<int>>{};
+
+    for (final id in event.messageIds) {
+      final message = cachedMessages[id] as StreamMessage?;
+      if (message == null) continue;
+      final StreamMessage(:streamId, :topic, :senderId, id: int messageId) = message;
+      (messagesByUserInStream[(streamId, senderId)] ??= []).add(messageId);
+      (messagesByUserInTopic[(streamId, topic, senderId)] ??= []).add(messageId);
+    }
+
+    for (final entry in messagesByUserInStream.entries) {
+      final MapEntry(key: (streamId, senderId), value: messages) = entry;
+      final sendersMap = streamSenders[streamId];
+      final idTracker = sendersMap?[senderId];
+
+      idTracker?.removeAll(messages);
+      if (idTracker?.maxId == null) sendersMap?.remove(senderId);
+      if (sendersMap?.isEmpty ?? false) streamSenders.remove(streamId);
+    }
+
+    for (final entry in messagesByUserInTopic.entries) {
+      final MapEntry(key: (streamId, topic, senderId), value: messages) = entry;
+      final topicsMap = topicSenders[streamId];
+      final sendersMap = topicsMap?[topic];
+      final idTracker = sendersMap?[senderId];
+
+      idTracker?.removeAll(messages);
+      if (idTracker?.maxId == null) sendersMap?.remove(senderId);
+      if (sendersMap?.isEmpty ?? false) topicsMap?.remove(topic);
+      if (topicsMap?.isEmpty ?? false) topicSenders.remove(streamId);
+    }
+  }
+}
+
+@visibleForTesting
+class MessageIdTracker {
+  /// A list of distinct message IDs, sorted ascendingly.
+  @visibleForTesting
+  QueueList<int> ids = QueueList.from([]);
+
+  /// The maximum id in the tracker list, or `null` if the list is empty.
+  int? get maxId => ids.lastOrNull;
+
+  /// Add the message ID to the tracker list at the proper place, if not present.
+  ///
+  /// Optimized, taking O(1) time for the case where that place is the end,
+  /// because that's the common case for a message that is received through
+  /// [PerAccountStore.handleEvent]. May take O(n) time in some rare cases.
+  void add(int id) {
+    final i = lowerBound(ids, id);
+    if (i < ids.length && ids[i] == id) {
+      // The ID is already present. Nothing to do.
+      return;
+    }
+    if (i == ids.length) {
+      ids.addLast(id);
+    } else {
+      ids.insert(i, id);
+    }
+  }
+
+  /// Add the messages IDs to the tracker list at the proper place, if not present.
+  ///
+  /// [newIds] should be sorted ascendingly.
+  void addAll(List<int> newIds) {
+    if (ids.isEmpty) {
+      ids.addAll(newIds);
+      return;
+    }
+    ids = setUnion(ids, newIds);
+  }
```

Comment:
This now takes $O(n + m)$ as `setUnion` gets rid of the shared IDs too in addition to keeping the resulting list in order.

Also, would it be better to replace `ids.addAll(newIds)` with `ids = newIds`? It should align it with the last line of the method.